okhttp/src/jvmMain/kotlin/okhttp3/internal/http2/Http2Connection.kt

```diff
@@ -237,38 +237,53 @@ class Http2Connection internal constructor(builder: Builder) : Closeable {
     val stream: Http2Stream
     val streamId: Int
 
-    synchronized(writer) {
-      synchronized(this) {
-        if (nextStreamId > Int.MAX_VALUE / 2) {
-          shutdown(REFUSED_STREAM)
-        }
-        if (isShutdown) {
-          throw ConnectionShutdownException()
+    // trigger interrupts before we write to the buffered output
+    // since flush is likely where we get notified.
+    if (Thread.currentThread().isInterrupted) {
+      // Throwing an IOException, so leave interrupt status.
+      throw InterruptedIOException()
+    }
+
+    try {
+      synchronized(writer) {
+        synchronized(this) {
+          if (nextStreamId > Int.MAX_VALUE / 2) {
+            shutdown(REFUSED_STREAM)
+          }
+          if (isShutdown) {
+            throw ConnectionShutdownException()
+          }
+          streamId = nextStreamId
+          nextStreamId += 2
+          stream = Http2Stream(streamId, this, outFinished, inFinished, null)
+          flushHeaders = !out ||
+            writeBytesTotal >= writeBytesMaximum ||
+            stream.writeBytesTotal >= stream.writeBytesMaximum
+          if (stream.isOpen) {
+            streams[streamId] = stream
+          }
         }
-        streamId = nextStreamId
-        nextStreamId += 2
-        stream = Http2Stream(streamId, this, outFinished, inFinished, null)
-        flushHeaders = !out ||
-          writeBytesTotal >= writeBytesMaximum ||
-          stream.writeBytesTotal >= stream.writeBytesMaximum
-        if (stream.isOpen) {
-          streams[streamId] = stream
+        if (associatedStreamId == 0) {
+          writer.headers(outFinished, streamId, requestHeaders)
+        } else {
+          require(!client) { "client streams shouldn't have associated stream IDs" }
+          // HTTP/2 has a PUSH_PROMISE frame.
+          writer.pushPromise(associatedStreamId, streamId, requestHeaders)
         }
       }
-      if (associatedStreamId == 0) {
-        writer.headers(outFinished, streamId, requestHeaders)
-      } else {
-        require(!client) { "client streams shouldn't have associated stream IDs" }
-        // HTTP/2 has a PUSH_PROMISE frame.
-        writer.pushPromise(associatedStreamId, streamId, requestHeaders)
+
+      if (flushHeaders) {
+        writer.flush()
       }
-    }
 
-    if (flushHeaders) {
-      writer.flush()
+      return stream
+    } catch (iioe: InterruptedIOException) {
```

Comment:
@swankjesse Alternatives here

a) If interrupted - don't call flush, instead write a RST_STREAM, and throw an InterruptedIOException. Keep connection healthy but generally make sure the 
b) Do this work via taskrunner so it can't be interrupted. Catch the interruption in say CountDownLatch in the caller thread, and then cancel the stream.